From 5a1d6e1417c19625eb0fcc19abb7e62d394dceca Mon Sep 17 00:00:00 2001
From: Christof Kaufmann <christof.kaufmann@hs-bochum.de>
Date: Tue, 2 May 2023 13:58:56 +0200
Subject: [PATCH] Skip Mujoco tests if OpenGL is not available

When selecting OSMesa by setting the environment variable
`MUJOCO_GL=osmesa`, this does not lead to an `ImportError` from Mujoco,
but to an `AttributeError` from OpenGL, if OSMesa is not available. This
commit adds `AttributeError` to the expected errors, when making the
test environments. It triggers a warning, but does not lead to a skipped
test. This behavior is analog to the case when EGL is selected, but not
available.

Additionally this commit tests if an OffScreenViewer object can be
instantiated. It fails when setting the environment variable
`MUJOCO_GL=glfw`, if GLFW is not available. In this case, the commit
skips the rendering tests with Mujoco environments.
---
 tests/envs/test_rendering.py | 26 +++++++++++++++++++++++++-
 tests/envs/utils.py          |  1 +
 2 files changed, 26 insertions(+), 1 deletion(-)

diff --git a/tests/envs/test_rendering.py b/tests/envs/test_rendering.py
index 2bec341c..a4887306 100644
--- a/tests/envs/test_rendering.py
+++ b/tests/envs/test_rendering.py
@@ -5,6 +5,25 @@ from gymnasium.logger import warn
 from tests.envs.utils import all_testing_env_specs
 
 
+try:
+    # raises an ImportError on egl and osmesa, if not available
+    import mujoco
+
+    from gymnasium.envs.mujoco.mujoco_rendering import OffScreenViewer
+
+    model = mujoco.MjModel.from_xml_string(
+        '<mujoco><worldbody><geom size="1"/></worldbody></mujoco>'
+    )
+    data = mujoco.MjData(model)
+
+    # raises a mujoco.FatalError on glfw, if not available
+    osv = OffScreenViewer(model, data)
+    osv.close()
+    skip_mujoco = False
+except:  # noqa: E722 (cannot catch mujoco.FatalError explicitly)
+    skip_mujoco = True
+
+
 def check_rendered(rendered_frame, mode: str):
     """Check that the rendered frame is as expected."""
     if mode == "rgb_array_list":
@@ -40,7 +59,12 @@ def check_rendered(rendered_frame, mode: str):
 
 # We do not check render_mode for some mujoco envs and any old Gym environment wrapped by `GymEnvironment`
 render_mode_env_specs = [
-    spec
+    pytest.param(
+        spec,
+        marks=pytest.mark.skipif(
+            skip_mujoco and "mujoco" in spec.entry_point, reason="OpenGL not available"
+        ),
+    )
     for spec in all_testing_env_specs
     if "mujoco" not in spec.entry_point or "v4" in spec.id
 ]
diff --git a/tests/envs/utils.py b/tests/envs/utils.py
index e643d22b..a16133d7 100644
--- a/tests/envs/utils.py
+++ b/tests/envs/utils.py
@@ -22,6 +22,7 @@ def try_make_env(env_spec: EnvSpec) -> Optional[gym.Env]:
             return env_spec.make(disable_env_checker=True).unwrapped
         except (
             ImportError,
+            AttributeError,
             gym.error.DependencyNotInstalled,
             gym.error.MissingArgument,
         ) as e:
-- 
2.40.0

